### PR TITLE
Issue #6615: Add support for Java 14 switch and yield

### PR DESCRIPTION
**Pull request by nmancus1**
Issue #6615: Add support for Java 14 switch and yield

This PR will introduce support for Java 14 switch expressions, and yield statements. 
_________________________
### [Relevant JLS grammar for switch expressions:](https://docs.oracle.com/javase/specs/jls/se13/preview/switch-expressions.html#jep354-15.28)
<ins>UnaryExpression:</ins>
    _PreIncrementExpression_
    _PreDecrementExpression_
     _`+` UnaryExpression_
     _`-` UnaryExpression_
    _UnaryExpressionNotPlusMinus_

<ins>PreIncrementExpression:</ins>
    _`++` UnaryExpression_

<ins>PreDecrementExpression:</ins>
    _`--` UnaryExpression_

<ins>UnaryExpressionNotPlusMinus:</ins>
_PostfixExpression_
_`~` UnaryExpression_
_`!` UnaryExpression_
_CastExpression_
**SwitchExpression**

<ins>**SwitchExpression:**</ins>
_`switch` ( Expression ) SwitchBlock_

<ins>SwitchBlock:</ins>
    _{ SwitchLabeledRule {SwitchLabeledRule} }_
    _{ {SwitchLabeledStatementGroup} {SwitchLabel :} }_

<ins>SwitchLabeledRule:</ins>
_SwitchLabeledExpression_
_SwitchLabeledBlock_
_SwitchLabeledThrowStatement_

<ins>SwitchLabeledExpression:</ins>
_SwitchLabel -> Expression ;_

<ins>SwitchLabeledBlock:</ins>
_SwitchLabel -> Block_

<ins>SwitchLabeledThrowStatement:</ins>
_SwitchLabel -> ThrowStatement_

<ins>SwitchLabel:</ins>
_`case` CaseConstant {, CaseConstant}_
_`default`_

<ins>CaseConstant:</ins>
_ConditionalExpression_

<ins>SwitchLabeledStatementGroup:</ins>
_SwitchLabel : {SwitchLabel :} BlockStatements_

_____________________________
### [Relevant grammar for yield statements:](https://docs.oracle.com/javase/specs/jls/se13/preview/switch-expressions.html#jep354-14.21)

<ins>YieldStatement:</ins>
_`yield` Expression ;_

_______________________________

I have introduced the new token, SWITCH_RULE,  for the following reasons:

1. From[ JLS](https://docs.oracle.com/javase/specs/jls/se13/preview/switch-expressions.html#jep354-14.11.1): _When a selector expression matches a switch label for a switch labeled rule, the labeled expression or statement is executed and nothing else._

    This means that there is no reason to check for fall-through, as in the case of switch case groups.  In checks, we can simply conclude that we are in a switch rule, and know that there will be definite assignment and no fall through. 

2. The java grammar differentiates between the two, so we should also. 

    **_The body of both a switch statement and a switch expression (15.28) is known as a switch block. This block can consist of either:_**

    _Switch labeled rules, which use -> to introduce either a switch labeled expression, switch labeled block, or a switch labeled throw statement; or_

    _Switch labeled statement groups, which use : to introduce switch labeled block statements._

### CODE REVIEW

**Comment by nmancus1** [comment]
This post is for issues.

**Comment by nmancus1** [comment]
This post is for reports

File: config/ant-phase-verify.xml [hunk 1]
```diff
@@ -157,6 +159,9 @@
          <exclude name="**/InputJava14Records.java"/>
          <!-- until https://github.com/checkstyle/checkstyle/issues/7103 -->
          <exclude name="**/InputJava14TextBlocks.java"/>
+         <!-- until https://github.com/checkstyle/checkstyle/issues/6615 -->
```
Comment (nmancus1) [inline config/ant-phase-verify.xml]: I need to delete this, but will it mess up CI since both commits are on here?

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g [hunk 2]
```diff
@@ -1244,6 +1268,36 @@ caseSList
         {#caseSList = #(#[SLIST,"SLIST"],#caseSList);}
     ;
 
+switchRule
+    :   (       (switchLabeledExpr)=>     se:switchLabeledExpr
+        |       (switchLabeledBlk)=>      sb:switchLabeledBlk
+        |       (switchLabeledThrow)=>    st:switchLabeledThrow
+        )
+        {## = #(#[CASE_GROUP, "CASE_GROUP"], se, sb, st);}
```
Comment (nmancus1) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:1276]: Should we re-use `CASE_GROUP`, as I've done here, or make a new token? Something like `SWITCH_RULE`?  Reusing CASE_GROUP will make check updates simpler, but would anything be gained by creating a new token?

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java [hunk 3]
```diff
@@ -3596,6 +3597,20 @@
     public static final int RECORD_DEF =
             GeneratedJavaTokenTypes.RECORD_DEF;
 
+    /**
+     * The {@code yield} keyword.  This element appears
```
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:3601]: Not directly related  to this line, but what should we do in the way of updating the switch statement documentation?  Have two different examples, then include a yield in each?

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g (hunk 2 above)
Comment (pbludov) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:1280]: This is a good question. I see at least a problem with [FallThrough](https://checkstyle.org/config_coding.html#FallThrough) check. The Java 14 switch expression cases may not contain a `break`, so almost all of them will be reported by this check. And it seems that writing the utility method `isItCaseGroupYouAreLookingFor` will be difficult. By adding a new token instead of case group this problem goes away. Another one appears - in some cases,`case_group` will be skipped, but it will be easier to fix, usually by simply adding one `logical or`.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g [hunk 4]
```diff
@@ -207,6 +208,15 @@ tokens {
...
     }
+
+    /**
+    * This int value tracks the depth of a switch expression. Along with the
+    * IDENT to id rule at the end of the parser, this value helps us
+    * to know if the "yield" we are parsing is an IDENT, method call, class,
+    * field, etc. or if it is a java 13+ yield statement. Positive values
+    * indicate that we are within a (possibly nested) switch expression.
+    */
+    private int switchRuleOrStatementDepth = 0;
```
Comment (esilkensen) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:219]: Would a name like "switchBlockDepth" make sense here?

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g (hunk 2 above)
Comment (esilkensen) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:1280]: Good callout @nmancus1, what you said makes sense to me @pbludov

Comment (nmancus1) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:1280]: @romani What do you think?

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g (hunk 4 above)
Comment (nmancus1) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:219]: Yes, that does make more sense.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g (hunk 2 above)
Comment (nmancus1) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g]: I have introduced the new token, SWITCH_RULE,  for the following reasons:

1. From[ JLS](https://docs.oracle.com/javase/specs/jls/se13/preview/switch-expressions.html#jep354-14.11.1): _When a selector expression matches a switch label for a switch labeled rule, the labeled expression or statement is executed and nothing else._

    This means that there is no reason to check for fall-through, as in the case of switch case groups.  In checks, we can simply conclude that we are in a switch rule, and know that there will be definite assignment and no fall through. 

2. The java grammar differentiates between the two, so we should also. 

    **_The body of both a switch statement and a switch expression (15.28) is known as a switch block. This block can consist of either:_**

    _Switch labeled rules, which use -> to introduce either a switch labeled expression, switch labeled block, or a switch labeled throw statement; or_

    _Switch labeled statement groups, which use : to introduce switch labeled block statements._

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g (hunk 4 above)
Comment (nmancus1) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g]: Fixed.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java [hunk 5]
```diff
@@ -3815,6 +3818,77 @@
...
+     * |   |   |--COMMA (,)
+     * |   |   `--EXPR
+     * |   |       `--IDENT (SUN)
+     * |   |--LAMBDA ({@code ->})
+     * |   |--EXPR
+     * |   |   `--STRING_LITERAL ("Weekend" )
+     * |   `--SEMI (;)
+     * |--SWITCH_RULE
+     * |   |--LITERAL_DEFAULT (default)
+     * |   |--LAMBDA {@code ->})
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:3872]: missing `(`

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g [hunk 6]
```diff
@@ -121,7 +121,7 @@ tokens {
     PATTERN_VARIABLE_DEF; RECORD_DEF; LITERAL_record="record";
     RECORD_COMPONENTS; RECORD_COMPONENT_DEF; COMPACT_CTOR_DEF;
     TEXT_BLOCK_LITERAL_BEGIN; TEXT_BLOCK_CONTENT;
-    TEXT_BLOCK_LITERAL_END;
+    TEXT_BLOCK_LITERAL_END; LITERAL_yield="yield"; SWITCH_RULE;
```
Comment (romani) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:124]: please move them both to new line, we try to keep  them groupped only if related.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.java [hunk 7]
```diff
@@ -1,12 +1,45 @@
...
 import java.util.Set;
+import java.util.function.IntConsumer;
 import java.util.function.Supplier;
 
 public class InputJava14SwitchExpression {
 
+    static class HardToParse {
+        void foo() {
+            Instant Ascii;
+            byte[] decodabet = new byte[1];
```
Comment (romani) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.java:18]: how this method code is related to switch ?

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java [hunk 8]
```diff
@@ -133,7 +133,7 @@
                         "PATTERN_VARIABLE_DEF", "RECORD_DEF", "LITERAL_RECORD",
                         "RECORD_COMPONENTS", "RECORD_COMPONENT_DEF", "COMPACT_CTOR_DEF",
                         "TEXT_BLOCK_LITERAL_BEGIN", "TEXT_BLOCK_CONTENT",
-                        "TEXT_BLOCK_LITERAL_END").collect(
+                        "TEXT_BLOCK_LITERAL_END", "LITERAL_YIELD", "SWITCH_RULE").collect(
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java:136]: place them both to new line.
It make sense to put `.collect(` to new line probably to reduce diff next time.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.java (hunk 7 above)
Comment (nmancus1) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.java:18]: I found this code during regression testing, line 19 and 25 caused exceptions during parsing when I was still struggling with `castExpression`  (along with the code on line 25). I put all code that gave problems in this static class to make sure that we can keep parsing it.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java (hunk 8 above)
Comment (nmancus1) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java]: Fixed.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g (hunk 6 above)
Comment (nmancus1) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g]: Fixed.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g [hunk 9]
```diff
@@ -1250,6 +1274,36 @@ caseSList
         {#caseSList = #(#[SLIST,"SLIST"],#caseSList);}
     ;
 
+switchRule
+    :   (       (switchLabeledExpr)=>     se:switchLabeledExpr
+        |       (switchLabeledBlk)=>      sb:switchLabeledBlk
```
[the remark this answers is not in the document]
Comment (romani) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:1279]: I agree, we have space, no need to be cryptic

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java [hunk 10]
```diff
@@ -3815,6 +3818,77 @@
...
+
+    /**
+     * A switch labeled rule has a switch label and either an expression,
+     * block, or throw statement. A switch labeled statement group has one
+     * or more switch labels and zero or more block statements. A switch label
+     * is either a case or default label.
+     *
+     * <p>For example:</p>
+     * <pre>
+     * switch (day) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:3846]: `return switch (day) {`

to make it clear that it is expression like `switch`, not a regular `switch`.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g (hunk 9 above)
[the remark this answers is not in the document]
Comment (nmancus1) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g]: Fixed.  I missed this item previously, apologies.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java (hunk 10 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java]: Fixed.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g [hunk 11]
```diff
@@ -1166,16 +1181,11 @@ traditionalStatement
...
-        |    "switch"^ LPAREN expression RPAREN LCURLY
-                ( casesGroup )*
-            RCURLY
-
         // exception try-catch block
         |    tryBlock
 
         // throw an exception
-        |    "throw"^ expression SEMI
+        |    throwStatement
```
Comment (romani) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:1188]: please move this update to separate commit as minor refactoring before functional update.
There will be 3 commits in this PR.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g [hunk 12]
```diff
@@ -1166,16 +1181,11 @@ traditionalStatement
         // Return an expression
         |    "return"^ (expression)? SEMI
 
-        // switch/case statement
-        |    "switch"^ LPAREN expression RPAREN LCURLY
-                ( casesGroup )*
-            RCURLY
-
```
Comment (romani) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:1173]: if movement to line 1139 was not done on good reason, please do this update in separate commit.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g [hunk 13]
```diff
@@ -1250,6 +1274,36 @@ caseSList
         {#caseSList = #(#[SLIST,"SLIST"],#caseSList);}
     ;
 
+switchRule
+    :   (       (switchLabeledExpression)=>     se:switchLabeledExpression
+        |       (switchLabeledBlock)=>      sb:switchLabeledBlock
+        |       (switchLabeledThrow)=>    st:switchLabeledThrow
```
Comment (romani) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:1280]: as we do this game of vertical code alignment: 
```
    :   (       (switchLabeledExpression)=>  se:switchLabeledExpression
        |       (switchLabeledBlock)=>           sb:switchLabeledBlock
        |       (switchLabeledThrow)=>          st:switchLabeledThrow
```

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java [hunk 14]
```diff
@@ -3815,6 +3818,80 @@
     public static final int TEXT_BLOCK_LITERAL_END =
             GeneratedJavaTokenTypes.TEXT_BLOCK_LITERAL_END;
 
+    /**
+     * The {@code yield} keyword.  This element appears
+     * as part of a yield statement.
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:3823]: we need AST example to show context of usage.
as `yield` is very specific "keyword" that is activated in switch only, in comparison to other keywords that do not allow to create any identifier with same name.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java [hunk 15]
```diff
@@ -3815,6 +3818,80 @@
...
+     * @since 8.36
+     **/
+    public static final int LITERAL_YIELD =
+            GeneratedJavaTokenTypes.LITERAL_yield;
+
+    /**
+     * A switch labeled rule has a switch label and either an expression,
+     * block, or throw statement. A switch labeled statement group has one
+     * or more switch labels and zero or more block statements. A switch label
+     * is either a case or default label.
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:3842]: wording is very strange.
It looks like brief of grammar. I doubt user need it, we never did this for other tokens.

Can we simply do ? 
`Switch Expressions` as content.
all details on what is possible should be read in java specification(that might change in future), this file is not good place for this.

Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java]: Fixed.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g (hunk 11 above)
Comment (nmancus1) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:1188]: Fixed.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g (hunk 12 above)
Comment (nmancus1) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:1173]: Fixed.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java (hunk 14 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:3823]: Fixed.

Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:3823]: Fixed.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g (hunk 13 above)
Comment (nmancus1) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g]: Fixed.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g [hunk 16]
```diff
@@ -1537,6 +1539,10 @@ castExpression
         )
     ;
 
+ switchExpression
+    :   "switch"^ LPAREN expression RPAREN switchBlock
```
Comment (romani) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:1598]: it should be 
```
"switch"^ LPAREN expression RPAREN LCURLY
                ( casesGroup )*
            RCURLY
```

as this commit do no functional changes

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java [hunk 17]
```diff
@@ -3815,6 +3818,123 @@
     public static final int TEXT_BLOCK_LITERAL_END =
             GeneratedJavaTokenTypes.TEXT_BLOCK_LITERAL_END;
 
+    /**
+     * The {@code yield} keyword.  This element appears
+     * as part of a yield statement.
+     *
+     * <p>For example:</p>
+     * <pre>
+     * return switch (mode) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:3827]:
```
int yield = 0;
return switch (mode) {
```

to show that yield is not literal out of switch.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g (hunk 16 above)
Comment (nmancus1) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:1598]: From [JLS](https://docs.oracle.com/javase/specs/jls/se13/preview/switch-expressions.html#jep354-15.28): 
<ins>**SwitchExpression:**</ins>
_`switch` ( Expression ) SwitchBlock_

Isn't it a good idea to follow JLS when possible?

Comment (romani) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:1598]: it is how it should looks in third commit, but in second commit it should be as it was before, commit is "minor: ...."

Comment (nmancus1) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:1598]: Oh, yes. I understand. Thanks for explaining that.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java (hunk 17 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:3828]: Fixed.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g (hunk 16 above)
Comment (nmancus1) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:1598]: Fixed.

**Comment by nmancus1** [comment]
Reviewers, all reports have been updated after latest changes. See https://github.com/checkstyle/checkstyle/pull/8449#issuecomment-658278229

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g (hunk 16 above)
Comment (romani) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:1598]: now in second commit, we have duplicates of 
```
"switch"^ LPAREN expression RPAREN LCURLY
                ( casesGroup )*
            RCURLY
```

please explain why you moved it up ?

Comment (nmancus1) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:1598]: I moved it above the `expression` rule to avoid some ambiguity.

Comment (romani) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:1598]: if such move up is required to new switch feature - it should stay in functional fix. 
But in second commit you can move keep moving in separate rule.

**Comment by nmancus1** [comment]
image

First and second commit pass mvn clean verify

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g (hunk 16 above)
Comment (nmancus1) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:1598]: Fixed.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java [hunk 18]
```diff
@@ -3815,6 +3818,133 @@
...
+     * @see #LITERAL_SWITCH
+     * @see #CASE_GROUP
+     * @see #SLIST
+     * @see #SWITCH_RULE
+     *
+     * @see <a href="https://docs.oracle.com/javase/specs/jls/se13/preview/switch-expressions.html">
+     * Java Language Specification, &sect;14.21</a>
+     *
+     * @since 8.36
+     **/
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:3890]: Ending javadoc usually only has 1 `*`.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java [hunk 19]
```diff
@@ -3815,6 +3818,133 @@
...
+     * @see #LITERAL_CASE
+     * @see #LITERAL_DEFAULT
+     * @see #LITERAL_SWITCH
+     * @see #LITERAL_YIELD
+     *
+     * @see <a href="https://docs.oracle.com/javase/specs/jls/se13/preview/switch-expressions.html">
+     * Java Language Specification, &sect;14.21</a>
+     *
+     * @since 8.36
+     **/
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:3944]: same.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java [hunk 20]
```diff
@@ -3815,6 +3818,133 @@
...
             GeneratedJavaTokenTypes.TEXT_BLOCK_LITERAL_END;
 
+    /**
+     * The {@code yield} keyword.  This element appears
+     * as part of a yield statement.
+     *
+     * <p>For example:</p>
+     * <pre>
+     * int yield = 0; // not a keyword here
+     * return switch (mode) {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:3828]: Do you have any inputs similar to this where we have a switch use a variable with the name `yield` to show your changes don't break existing parser with yield variables?

**Comment by rnveach** [comment]
Has there been any time regression done for antlr changes to see if changes had any negative impact on time taken to parse?

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.java [hunk 21]
```diff
@@ -1,12 +1,45 @@
...
+        class yieldClass{
+            int yield = 6;
+
+            public void yield(){
+
+            }
+        }
+
+    }
+
```
Comment (nmancus1) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.java:42]: @rnveach

**Comment by nmancus1** [comment]
>Has there been any time regression done for antlr changes to see if changes had any negative impact on time taken to parse?

Not since full records support.   Should I run trials against 8.34?

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java (hunk 20 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:3828]: See https://github.com/checkstyle/checkstyle/pull/8449#discussion_r467665802

**Comment by romani** [comment]
Let's move performance consideration to separate issue. It should not block ability to parse in general.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.java (hunk 21 above)
Comment (rnveach) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.java:42]: Based on the grammar changes, you would need the `yield` usage in the `switch` and not just in random places, imo.

Comment (nmancus1) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.java:42]: That will not compile in jdk14.

File: src/test/resources/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.txt [hunk 22]
```diff
@@ -0,0 +1,3266 @@
...
+    |       |   |       |--CASE_GROUP -> CASE_GROUP [492:12]
+    |       |   |       |   |--LITERAL_DEFAULT -> default [492:12]
+    |       |   |       |   |   `--COLON -> : [492:19]
+    |       |   |       |   `--SLIST -> SLIST [493:16]
+    |       |   |       |       `--LITERAL_YIELD -> yield [493:16]
+    |       |   |       |           |--EXPR -> EXPR [493:22]
+    |       |   |       |           |   `--UNARY_MINUS -> - [493:22]
+    |       |   |       |           |       `--NUM_INT -> 1 [493:24]
+    |       |   |       |           `--SEMI -> ; [493:25]
+    |       |   |       `--RCURLY -> } [494:8]
```
Comment (nmancus1) [inline src/test/resources/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.txt:3263]: Grammar correctly parses "yield" idenfitfier usage within switch.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.java [hunk 23]
```diff
@@ -449,4 +482,15 @@ private String expression2(T t) {
...
+    private static int getValueViaYield(String mode) {
+        int yield = 0;
+        return switch (mode) {
+            case "a", "b":
+                int x = yield + 2;
+                yield x;
+            default:
+                yield - 1;
+        };
+    }
```
Comment (nmancus1) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.java:495]: @rnveach Please see https://github.com/checkstyle/checkstyle/pull/8449/files#r468609366

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.java (hunk 21 above)
Comment (nmancus1) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.java:42]: https://github.com/checkstyle/checkstyle/pull/8449/files#r468609715

File: src/test/resources/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.txt [hunk 24]
```diff
@@ -0,0 +1,3324 @@
...
+    |       |   |       |       |   |   `--LITERAL_INT -> int [500:16]
+    |       |   |       |       |   |--IDENT -> yield [500:20]
+    |       |   |       |       |   `--ASSIGN -> = [500:26]
+    |       |   |       |       |       `--EXPR -> EXPR [500:28]
+    |       |   |       |       |           `--NUM_INT -> 42 [500:28]
+    |       |   |       |       |--SEMI -> ; [500:30]
+    |       |   |       |       `--LITERAL_YIELD -> yield [501:16]
+    |       |   |       |           |--EXPR -> EXPR [501:22]
+    |       |   |       |           |   `--IDENT -> yield [501:22]
+    |       |   |       |           `--SEMI -> ; [501:27]
```
Comment (nmancus1) [inline src/test/resources/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.txt:3311]: Correctly parsed here, too.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.java [hunk 25]
```diff
@@ -449,4 +482,25 @@ private String expression2(T t) {
...
+
+    private static int getValueViaYield2(String mode) {
+        return switch (mode) {
+            case "a", "b":
+                int yield = 42;
+                yield yield;
+            default:
+                yield - 1;
+        };
+    }
```
Comment (nmancus1) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.java:505]: @rnveach here also: https://github.com/checkstyle/checkstyle/pull/8449/files#r468615659

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.java (hunk 21 above)
Comment (nmancus1) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14SwitchExpression.java:42]: And here: https://github.com/checkstyle/checkstyle/pull/8449/files#r468615917

**Comment by rnveach** [comment]
Can be merged when CI passes.

**Comment by nmancus1** [comment]
> Has there been any time regression done for antlr changes to see if changes had any negative impact on time taken to parse?

Before my changes:
image


After my changes:
image
